src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/Ordering/OrderingHelper.cs

```diff
@@ -168,6 +168,63 @@ public static IProjectTree GetFirstChild(IProjectTree projectTree)
             return GetChildren(projectTree).FirstOrDefault();
         }
 
+        /// <summary>
+        /// Get the difference to see what was added.
+        /// We do a sanity check to make sure they have a valid display order.
+        /// We also order the added nodes by their display order.
+        /// </summary>
+        public static IEnumerable<IProjectTree> GetAddedNodes(IProjectTree previousTarget, IProjectTree currentTarget)
+        {
+            return
+                currentTarget.Children.Where(updatedChild => {
+                    var hasChild = previousTarget.TryFind(updatedChild.Identity, out var child);
+                    var hasValidUpdatedChild = HasValidDisplayOrder(updatedChild);
+
+                    return (!hasChild || (hasChild && !HasValidDisplayOrder(child))) && hasValidUpdatedChild;
+                })
+                .OrderBy(GetDisplayOrder).ToList();
+        }
+
+        /// <summary>
+        /// This moves the nodes above the target's first child. 
+        /// If there are no children of the target and the target is an empty folder, it will try the parent target's first child.
+        /// </summary>
+        public static void MoveNodesToTop(Project project, IEnumerable<IProjectTree> nodes, IProjectTree target)
+        {
+            // Move added nodes to the top. 
+            var nodeToAddAbove = GetFirstChild(target);
+
+            // This could have been an empty folder and nodeToAddAbove is null, therefore
+            //     get the first child of the target's parent.
+            // Empty folders do not have a valid display order.
+            // In the future, if we want to order empty folders, meaning they have a valid display order, 
+            //     this will need to be extended to handle that behavior.
+            if (nodeToAddAbove == null && target.IsFolder && !target.Children.Any())
+            {
+                nodeToAddAbove = GetFirstChild(target.Parent);
+            }
+
+            if (nodeToAddAbove != null)
+            {
+                foreach (var addedNode in nodes)
+                {
+                    TryMoveAbove(project, addedNode, nodeToAddAbove);
+                }
+            }
+        }
+
+        /// <summary>
+        /// This moves the nodes above the target's first child. 
+        /// If there are no children of the target and the target is an empty folder, it will try the parent target's first child.
+        /// </summary>
+        public static async Task MoveNodesToTopAsync(ConfiguredProject configuredProject, IEnumerable<IProjectTree> nodes, IProjectTree target)
+        {
+            await ModifyProjectAsync(configuredProject, project => {
+                MoveNodesToTop(project, nodes, target);
+                return true;
+            }).ConfigureAwait(true);
```

Comment:
Why "true" here? Do you need to switch back to original thread if you ended up switching threads?

Reply:
I'm going to rename this to TryMoveNodes*. I will also make it return true if all nodes were successfully moved to the right place. If one node was not moved successfully, then it will return false.

Reply:
I mean, why `ConfigureAwait(true)`? We use `ConfigureAwait(false)` to avoid unnecessarily thread switches unless absolutely needed.

Reply:
why is this method `async` at all?  Why not just return the task to the caller?
```C#
public static Task MoveNodesToTopAsync(ConfiguredProject configuredProject, IEnumerable<IProjectTree> nodes, IProjectTree target)
{
    return ModifyProjectAsync(configuredProject, project =>
    {
        MoveNodesToTop(project, nodes, target);
        return true;
    });
}
```

Reply:
The changes just return the task.